extensions/monzo/src/lib/monzo.ts

```diff
@@ -0,0 +1,113 @@
+  oauthClientId: string;
+  oauthClientSecret: string;
+}
+
+export async function getClient(): Promise<IMonzoClient> {
+  const tokenSet = await client.getTokens();
+
+  // If no access token, auth and try again.
+  if (!tokenSet || !tokenSet.accessToken) {
+    await initAuth();
+    return getClient();
+  }
+
+  // If needs refresh, refresh and try again
+  if (tokenSet.refreshToken && tokenSet.isExpired()) {
+    await client.setTokens(await refreshTokens(tokenSet?.refreshToken));
+    return getClient();
+  }
+
+  // If no refresh token, something has gone wrong
+  if (!tokenSet.refreshToken) {
+    throw new Error("No refresh token");
+  }
+
+  return MonzoClient(tokenSet.accessToken);
+}
+
+async function initAuth() {
+  const preferences = getPreferenceValues<Preferences>();
+  const authRequest = await client.authorizationRequest({
+    endpoint: authorizeUrl,
+    clientId: preferences.oauthClientId.trim(),
+    scope: "",
+    extraParameters: { x_authorize_url: monzoAuthorizeUri },
+  });
+  const { authorizationCode } = await client.authorize(authRequest);
+  const tokens = await fetchTokens(authRequest, authorizationCode);
+  await client.setTokens(tokens);
+}
+
+async function fetchTokens(
+  authRequest: OAuth.AuthorizationRequest,
+  authorizationCode: string
+): Promise<OAuth.TokenResponse> {
+  const preferences = getPreferenceValues<Preferences>();
+  const data = new URLSearchParams();
+  data.append("client_id", preferences.oauthClientId.trim());
+  data.append("code", authorizationCode);
+  data.append("code_verifier", authRequest.codeVerifier);
+  data.append("grant_type", "authorization_code");
+  data.append("redirect_uri", authRequest.redirectURI);
+
+  data.append("x_client_secret", preferences.oauthClientSecret.trim());
+  data.append("x_access_token_uri", monzoAccessTokenUri);
+
+  const response = await fetch(tokenUrl, {
+    method: "POST",
+    body: data,
+  });
+
+  if (!response.ok) {
+    console.error("Error completing sign-in:", await response.text());
+    throw new Error(response.statusText);
+  }
+
+  return (await response.json()) as OAuth.TokenResponse;
+}
+
+async function refreshTokens(
+  refreshToken: string
+): Promise<OAuth.TokenResponse> {
+  const preferences = getPreferenceValues<Preferences>();
+
+  const data = new URLSearchParams();
+  data.append("client_id", preferences.oauthClientId.trim());
+  data.append("refresh_token", refreshToken);
+  data.append("grant_type", "refresh_token");
+  data.append("client_secret", preferences.oauthClientSecret.trim());
+
+  const response = await fetch(monzoAccessTokenUri, {
+    method: "POST",
+    body: data,
+  });
+
+  if (!response.ok) {
+    console.error("Error refreshing access:", await response.text());
+    throw new Error(response.statusText);
+  }
+
+  const tokenResponse = (await response.json()) as OAuth.TokenResponse;
+  tokenResponse.refresh_token = tokenResponse.refresh_token ?? refreshToken;
+  return tokenResponse;
+}
```

Comment:
(I couldn't test it since I don't have a monzo account anymore. I only had one until they changed their name from Mondo to Monzo. Ah that pink alpha card 😭)

Reply:
Sorry for the late reply, I'll try this out at the weekend.